### PR TITLE
Add "-f" to the /move command

### PR DESCRIPTION
**Pull request by Daenges**
As a solution for #1872 I have implemented a `-f` argument, which forces the movement regardless of its potential consequences.

### CODE REVIEW

**Review by ORelio — CHANGES_REQUESTED** [review]
Looks good! However, please document that the switch exists in command help or error message when path is unsafe, otherwise nobody will know about it.

**Comment by Daenges** [comment]
@ORelio Changes applied.

File: MinecraftClient/Commands/Move.cs [hunk 1]
```diff
@@ -1,38 +1,36 @@
...
 using MinecraftClient.Mapping;
 
 namespace MinecraftClient.Commands
 {
     public class Move : Command
     {
         public override string CmdName { get { return "move"; } }
-        public override string CmdUsage { get { return "move <on|off|get|up|down|east|west|north|south|x y z>"; } }
-        public override string CmdDesc { get { return "walk or start walking."; } }
+        public override string CmdUsage { get { return "move <on|off|get|up (-f)|down (-f)|east (-f)|west (-f)|north (-f)|south (-f)|x y z>"; } }
```
Comment (ReinforceZwei) [inline MinecraftClient/Commands/Move.cs:11]: I think the help message here became too complex and hard to read. 

How about:
```suggestion
        public override string CmdUsage { get { return "move <on|off|get|up|down|east|west|north|south|x y z> [-f]"; } }
```

File: MinecraftClient/Commands/Move.cs [hunk 2]
```diff
@@ -1,38 +1,36 @@
...
+        public override string CmdDesc { get { return "walk or start walking. Included protection can be bypassed with \"-f\"."; } }
```
Comment (ReinforceZwei) [inline MinecraftClient/Commands/Move.cs:12]:
```suggestion
        public override string CmdDesc { get { return "walk or start walking. Unsafe path protection can be bypassed with \"-f\"."; } }
```

File: MinecraftClient/Commands/Move.cs [hunk 3]
```diff
@@ -1,38 +1,36 @@
...
 
         public override string Run(McClient handler, string command, Dictionary<string, object> localVars)
         {
-            string[] args = getArgs(command);
-            string argStr = getArg(command).Trim().ToLower();
+            string[] args = getArgs(command.ToLower());
```
Comment (ReinforceZwei) [inline MinecraftClient/Commands/Move.cs:16]: I think the parsing of the arguments could be done like below:
https://github.com/MCCTeam/Minecraft-Console-Client/blob/259ef80cf97b18a55da0f4b810460e1e49e1711a/MinecraftClient/Program.cs#L63-L71

We look for `-f` in the whole argument string, if found, set a boolean flag to true and remove the `-f` from the argument string. Then the position of `-f` will no longer be important, and could be easily apply the flag to moving with absolute coordinate.

**Review by ReinforceZwei — COMMENTED** [review]
How about also applying the `force` option for moving to absolute coordinate?

**Comment by Daenges** [comment]
> How about also applying the `-f` option for moving with absolute coordinate?

@ReinforceZwei 
Also thought about that. The only problem here is, that the absolute movement consists of several individual movement desicions. I do not think that players can actually estimate what happens, if the client goes on full risk on every movement. This can result in major damage, since the behaviour of the client on the way can result in unpredictable dangerous situations. (Especially problematic since you can not stop the movement of the client without disconnecting.)

My idea is that the player lets the bot autowalk **near** a dangerous situation and then switches to direct control. Thereby we force the user to be attentive and to be responsible of their own decisions.

**Comment by Daenges** [comment]
Despite my doubts I implemented the changes. Lets pray for the peoples common sense.

**Comment by ReinforceZwei** [comment]
That could be a problem for those who don't pay attention to the situation. Maybe forcing people to type the full `-force` argument could raise some attention? 

Never mind if you don't like unleashing unsafe path for absolute coordinate. Feel free to revert it back.

**Comment by Daenges** [comment]
I thought about git and how it also allows the -f command without any further barricades.

As long as people use the regular command, they get warned by a message and everything further they do is on their own responsibility.

Why restrict their freedom if they know what they want to do? Some need to touch the fire to know that it is hot.

File: MinecraftClient/Commands/Move.cs [hunk 4]
```diff
@@ -1,38 +1,43 @@
...
 
 namespace MinecraftClient.Commands
 {
     public class Move : Command
     {
         public override string CmdName { get { return "move"; } }
-        public override string CmdUsage { get { return "move <on|off|get|up|down|east|west|north|south|x y z>"; } }
-        public override string CmdDesc { get { return "walk or start walking."; } }
+        public override string CmdUsage { get { return "move <on|off|get|up|down|east|west|north|south|x y z> [-f]"; } }
+        public override string CmdDesc { get { return "walk or start walking. Included protection can be bypassed with \"-f\"."; } }
```
Comment (ORelio) [inline MinecraftClient/Commands/Move.cs:12]:
```suggestion
        public override string CmdDesc { get { return "walk or start walking. \"-f\": force unsafe movements like falling or touching fire"; } }
```

File: MinecraftClient/Commands/Move.cs [hunk 5]
```diff
@@ -41,26 +46,28 @@ public override string Run(McClient handler, string command, Dictionary<string,
...
                         int x = int.Parse(args[0]);
                         int y = int.Parse(args[1]);
                         int z = int.Parse(args[2]);
                         Location goal = new Location(x, y, z);
-                        if (handler.MoveTo(goal))
+
+                        if (handler.MoveTo(goal, allowUnsafe: takeRisk))
                             return Translations.Get("cmd.move.walk", goal);
-                        return Translations.Get("cmd.move.fail", goal);
+                        else return !takeRisk ? Translations.Get("cmd.move.suggestforce") : Translations.Get("cmd.move.fail", goal);
```
Comment (ORelio) [inline MinecraftClient/Commands/Move.cs:70]: How about removing the negation in ternary condition and also include coordinates:
```suggestion
                        else return takeRisk ? Translations.Get("cmd.move.fail", goal) : Translations.Get("cmd.move.suggestforce", goal);
```
(see below for coordinates in message)

File: MinecraftClient/Resources/lang/en.ini [hunk 6]
```diff
@@ -310,6 +310,7 @@ cmd.move.moving=Moving {0}
 cmd.move.dir_fail=Cannot move in that direction.
 cmd.move.walk=Walking to {0}
 cmd.move.fail=Failed to compute path to {0}
+cmd.move.suggestforce=The implemented protection prevents a movement in this direction. Use the -f parameter to force the movement.
```
Comment (ORelio) [inline MinecraftClient/Resources/lang/en.ini:313]: The path either does not exists _or_ is unsafe. So the message could look like this instead:
```suggestion
cmd.move.suggestforce=Failed to compute a safe path to {0}. Try -f parameter to allow unsafe movements.
```

File: MinecraftClient/Resources/lang/de.ini [hunk 7]
```diff
@@ -310,6 +310,7 @@ cmd.move.moving=Laufe {0}
 cmd.move.dir_fail=Kann nicht in diese Richtung laufen.
 cmd.move.walk=Gehe nach {0}
 cmd.move.fail=Konnte Pfad nach {0} nicht berechnen.
+cmd.move.suggestforce=Die implementierte Absicherung verhindert eine Bewegung in diese Richtung. Du kannst sie mit dem -f Parameter erzwingen.
```
Comment (ORelio) [inline MinecraftClient/Resources/lang/de.ini:313]: I cannot make suggestions in German but see the English version below.

**Review by ORelio — COMMENTED** [review]
Looks good to me! Here are a few additional suggestions mainly to improve wording/readability:

**Review by ORelio — APPROVED** [review]

**Comment by ORelio** [comment]
Good job! I'll leave @ReinforceZwei review again just in case and it should be good to merge 👍

File: MinecraftClient/Commands/Move.cs [hunk 8]
```diff
@@ -41,26 +46,28 @@ public override string Run(McClient handler, string command, Dictionary<string,
...
-                        default: return Translations.Get("cmd.look.unknown", argStr);
+                        default: return Translations.Get("cmd.look.unknown", args[0]);
                     }
                     if (Movement.CanMove(handler.GetWorld(), handler.GetCurrentLocation(), direction))
                     {
-                        handler.MoveTo(Movement.Move(handler.GetCurrentLocation(), direction));
-                        return Translations.Get("cmd.move.moving", argStr);
+                        if (handler.MoveTo(Movement.Move(handler.GetCurrentLocation(), direction), allowUnsafe: takeRisk))
+                            return Translations.Get("cmd.move.moving", args[0]);
+                        else return !takeRisk ? Translations.Get("cmd.move.suggestforce") : Translations.Get("cmd.move.dir_fail");
```
Comment (ReinforceZwei) [inline MinecraftClient/Commands/Move.cs:55]:
```suggestion
                        else return takeRisk ? Translations.Get("cmd.move.dir_fail") : Translations.Get("cmd.move.suggestforce");
```

**Pull request merged**